File: src/main/ipc/image-history.ts
import { ipcMain } from 'electron'
import * as fs from 'fs'

// In-memory undo/redo stacks per file
const undoStacks = new Map<string, Buffer[]>()
const redoStacks = new Map<string, Buffer[]>()

const MAX_HISTORY = 20

export function registerImageHistoryHandlers(): void {
  // Save current file state before an edit
  ipcMain.handle('history:beforeEdit', async (_e, filePath: string) => {
    const buffer = fs.readFileSync(filePath)
    const key = filePath.toLowerCase()

    if (!undoStacks.has(key)) undoStacks.set(key, [])
    const stack = undoStacks.get(key)!
    stack.push(buffer)
    if (stack.length > MAX_HISTORY) stack.shift()

    // Clear redo on new edit
    redoStacks.set(key, [])
    return { success: true }
  })

  ipcMain.handle('history:undo', async (_e, filePath: string) => {
    const key = filePath.toLowerCase()
    const undoStack = undoStacks.get(key)
    if (!undoStack || undoStack.length === 0) return { success: false }

    // Save current state to redo
    const current = fs.readFileSync(filePath)
    if (!redoStacks.has(key)) redoStacks.set(key, [])
    redoStacks.get(key)!.push(current)

    // Restore previous state
    const prev = undoStack.pop()!
    fs.writeFileSync(filePath, prev)
    return { success: true }
  })

  ipcMain.handle('history:redo', async (_e, filePath: string) => {
    const key = filePath.toLowerCase()
    const redoStack = redoStacks.get(key)
    if (!redoStack || redoStack.length === 0) return { success: false }

    // Save current state to undo
    const current = fs.readFileSync(filePath)
    if (!undoStacks.has(key)) undoStacks.set(key, [])
    undoStacks.get(key)!.push(current)

    // Restore redo state
    const next = redoStack.pop()!
    fs.writeFileSync(filePath, next)
    return { success: true }
  })
}

File: tauri2/src-tauri/src/commands/image_history.rs
use std::collections::HashMap;
use std::fs;
use std::sync::Mutex;

const MAX_HISTORY: usize = 20;

pub struct HistoryState {
    pub undo_stacks: HashMap<String, Vec<Vec<u8>>>,
    pub redo_stacks: HashMap<String, Vec<Vec<u8>>>,
}

impl Default for HistoryState {
    fn default() -> Self {
        Self {
            undo_stacks: HashMap::new(),
            redo_stacks: HashMap::new(),
        }
    }
}

#[tauri::command]
pub fn history_before_edit(
    file_path: String,
    state: tauri::State<'_, Mutex<HistoryState>>,
) -> Result<(), String> {
    let data = fs::read(&file_path).map_err(|e| e.to_string())?;
    let key = file_path.to_lowercase();
    let mut state = state.lock().map_err(|e| e.to_string())?;

    let stack = state.undo_stacks.entry(key.clone()).or_default();
    stack.push(data);
    if stack.len() > MAX_HISTORY {
        stack.remove(0);
    }

    state.redo_stacks.insert(key, Vec::new());
    Ok(())
}

#[tauri::command]
pub fn history_undo(
    file_path: String,
    state: tauri::State<'_, Mutex<HistoryState>>,
) -> Result<bool, String> {
    let key = file_path.to_lowercase();
    let mut state = state.lock().map_err(|e| e.to_string())?;

    // Pop from undo stack
    let prev = match state.undo_stacks.get_mut(&key) {
        Some(s) if !s.is_empty() => s.pop().unwrap(),
        _ => return Ok(false),
    };

    // Save current to redo
    let current = fs::read(&file_path).map_err(|e| e.to_string())?;
    state.redo_stacks.entry(key).or_default().push(current);

    // Restore
    fs::write(&file_path, prev).map_err(|e| e.to_string())?;
    Ok(true)
}

#[tauri::command]
pub fn history_redo(
    file_path: String,
    state: tauri::State<'_, Mutex<HistoryState>>,
) -> Result<bool, String> {
    let key = file_path.to_lowercase();
    let mut state = state.lock().map_err(|e| e.to_string())?;

    // Pop from redo stack
    let next = match state.redo_stacks.get_mut(&key) {
        Some(s) if !s.is_empty() => s.pop().unwrap(),
        _ => return Ok(false),
    };

    // Save current to undo
    let current = fs::read(&file_path).map_err(|e| e.to_string())?;
    state.undo_stacks.entry(key).or_default().push(current);

    // Restore
    fs::write(&file_path, next).map_err(|e| e.to_string())?;
    Ok(true)
}
